source/Timer.cpp

```diff
@@ -0,0 +1,322 @@
+/* Timer.cpp
+Copyright (c) 2023 by Timothy Collett
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "Timer.h"
+
+#include "DataNode.h"
+#include "DataWriter.h"
+#include "GameData.h"
+#include "Logger.h"
+#include "Planet.h"
+#include "PlayerInfo.h"
+#include "Random.h"
+#include "Ship.h"
+#include "StellarObject.h"
+#include "UI.h"
+
+using namespace std;
+
+Timer::Timer(const DataNode &node, const Mission *mission)
+{
+	Load(node, mission);
+}
+
+
+
+void Timer::Load(const DataNode &node, const Mission *mission)
+{
+	if(node.Size() > 1)
+		name = node.Token(1);
+
+	this->mission = mission;
+
+	for(const DataNode &child : node)
+	{
+		if(child.Token(0) == "time" && child.Size() > 1)
+		{
+			base = static_cast<int64_t>(child.Value(1));
+			if(child.Size() > 2)
+				rand = static_cast<uint32_t>(child.Value(2));
+
+			timeToWait = base + Random::Int(rand);
+		}
+		else if(child.Token(0) == "idle")
+			requireIdle = true;
+		else if(child.Token(0) == "optional")
+			optional = true;
+		else if(child.Token(0) == "system" && child.Size() > 1)
+			system = GameData::Systems().Get(child.Token(1));
+		else if(child.Token(0) == "system")
+			systems.Load(child);
+		else if(child.Token(0) == "proximity")
+		{
+			if(child.Size() > 1)
+				proximity = child.Value(1);
+			if(child.Size() > 2)
+			{
+				if(child.Token(2) == "close")
+					closeTo = true;
+				else if(child.Token(2) == "far")
+					closeTo = false;
+			}
+			if(child.HasChildren())
+			{
+				const DataNode &firstGrand = (*child.begin());
+				proximityCenter = GameData::Planets().Find(firstGrand.Token(0));
+				// If this is null, then it's not a planet by name, it's a LocationFiler
+				if(proximityCenter == nullptr)
+					proximityCenters.Load(firstGrand);
```

Comment:
Not a fan of first checking if the first token is a planet and loading the first node as a location filter if the first token isn't recognized as a planet. Even if this was fine, this won't work, as a planet may not be loaded yet by the time this mission is loading, meaning we could falsely assume that the token we were given is for a planet.

The syntax should probably look more something like this:
```html
proximity [<planet name>]
	[<location filter>]
"proximity settings" <distance#> [close | far]
```

That way the handling of `proximity` can be similar to the handling of `source` or `destination` in Mission, or of `system` in this PR.

Reply:
Ah, I forgot that loading order is not guaranteed. I'll rework that.